### PR TITLE
updated synonymGroups table-stream with tests

### LINKED ISSUE #2504: Update table-stream for new synonymGroups
Update the table-stream for the new design of synonymGroups for opensearch. 

As this cannot be tested locally, include testing of adding and removing records via this table-stream.

### PR DESCRIPTION
**Pull request by Courey**
# Description
This change is to modify the synonymGroups table-stream to match the format we decided at our in person team meeting. It includes full test coverage. 

# Related Issue(s)
Resolves #2504 

# Testing
Table-streams are not able to run locally because the architect sandbox uses dynalite which does not support table-streams. So this has not been run. However, there are 3 tests to cover each of the three actions possible, MODIFY, INSERT, and REMOVE. 

I would appreciate someone with dev access helping me test this on the dev server to ensure that it works properly.

### CODE REVIEW

**Review by lpsinger — CHANGES_REQUESTED** [review]
I think that there are some logical errors here. For example, your handler for removal of a DynamoDB record deletes the item in the OpenSearch index. However, it ought to _update_ the item in OpenSearch with the remaining event names that still correspond to the same UUID.

File: app/table-streams/synonyms/index.ts [hunk 1]
```diff
@@ -12,12 +12,14 @@ import { errors } from '@opensearch-project/opensearch'
...
+  StreamRecord,
 } from 'aws-lambda'
 
 import { createTriggerHandler } from '~/lib/lambdaTrigger.server'
-import type { Synonym } from '~/routes/synonyms/synonyms.lib'
+import type { Synonym, SynonymGroup } from '~/routes/synonyms/synonyms.lib'
+import { getSynonymsByUuid } from '~/routes/synonyms/synonyms.server'
 
-const index = 'synonyms'
+const index = 'synonymGroups'
```
[the remark this answers is not in the document]
Comment (Courey) [reply app/table-streams/synonyms/index.ts:22]: well it's not a synonym. those are structurally different than what we are putting in opensearch. I didn't want anyone to be confused by that.

**Review by lpsinger — COMMENTED** [review]
Would you please address the remaining outstanding comments from my first review?

File: app/table-streams/synonyms/index.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (lpsinger) [reply app/table-streams/synonyms/index.ts:21]: Fair enough.

File: app/table-streams/synonyms/index.ts [hunk 2]
```diff
@@ -34,26 +34,59 @@ async function removeIndex(id: string) {
...
 export const handler = createTriggerHandler(
   async ({ eventName, dynamodb }: DynamoDBRecord) => {
-    const id = unmarshallTrigger(dynamodb!.Keys).id as string
+    if (!eventName || !dynamodb) return
     const promises = []
-
-    if (eventName === 'REMOVE') {
+    const id = unmarshallTrigger(dynamodb!.Keys).synonymId
+    const synonym = unmarshallTrigger(dynamodb!.NewImage) as Synonym
+    const existingGroup = await getIndex(synonym)
```
Comment (lpsinger) [inline app/table-streams/synonyms/index.ts:66]: Remember that we are treating DynamoDB as the source of truth, not OpenSearch. Please get the list of synonyms matching the UUID of the altered record using DynamoDB.

File: app/table-streams/synonyms/index.ts [hunk 3]
```diff
@@ -34,26 +34,59 @@ async function removeIndex(id: string) {
...
+        eventIds: group.eventIds.filter((id) => id !== synonym.eventId),
+      } as SynonymGroup
+      promises.push(putIndex(modifiedGroup))
     } /* (eventName === 'INSERT' || eventName === 'MODIFY') */ else {
-      const synonym = unmarshallTrigger(dynamodb!.NewImage) as Synonym
-      promises.push(putIndex(synonym))
+      if (!group.eventIds.includes(synonym.eventId)) {
+        group.eventIds.push(synonym.eventId)
+      }
+      promises.push(putIndex(group as SynonymGroup))
```
Comment (lpsinger) [inline app/table-streams/synonyms/index.ts:88]: As I said the other day, you shouldn't need to compute the terms to add or remove if you just query all of the terms that match the UUID of the DynamoDB record that triggered the Lambda.

File: sandbox-search.js [hunk 4]
```diff
@@ -11,13 +11,28 @@ export default async function () {
...
   const { circulars, synonyms } = JSON.parse(text)
 
+  const groupedSynonyms = {}
+  synonyms.forEach((synonym) => {
+    if (groupedSynonyms[synonym.synonymId]) {
+      groupedSynonyms[synonym.synonymId].push(synonym.eventId)
+    } else {
+      groupedSynonyms[synonym.synonymId] = [synonym.eventId]
+    }
+  })
```
Comment (lpsinger) [inline sandbox-search.js:21]: Is there a builtin JavaScript method or a Lodash helper to group an iterable by the value of a key function?

**Review by lpsinger — CHANGES_REQUESTED** [review]

File: app/table-streams/synonyms/index.ts [hunk 5]
```diff
@@ -34,26 +35,31 @@ async function removeIndex(id: string) {
...
 
 export const handler = createTriggerHandler(
   async ({ eventName, dynamodb }: DynamoDBRecord) => {
-    const id = unmarshallTrigger(dynamodb!.Keys).id as string
+    if (!eventName || !dynamodb) return
     const promises = []
-
-    if (eventName === 'REMOVE') {
+    const id = unmarshallTrigger(dynamodb!.Keys).synonymId
+    const synonym = unmarshallTrigger(dynamodb!.NewImage) as Synonym
```
Comment (lpsinger) [inline app/table-streams/synonyms/index.ts:52]:
```suggestion
    const { synonymId } = unmarshallTrigger(dynamodb!.NewImage) as Synonym
```

File: app/table-streams/synonyms/index.ts [hunk 6]
```diff
@@ -34,26 +35,31 @@ async function removeIndex(id: string) {
...
+    const group = {
+      synonymId: synonym.synonymId,
+      eventIds: dynamoSynonyms.map((synonym) => synonym.eventId),
+    }
+    if (eventName === 'REMOVE' && group.eventIds.length === 0) {
       promises.push(removeIndex(id))
     } /* (eventName === 'INSERT' || eventName === 'MODIFY') */ else {
-      const synonym = unmarshallTrigger(dynamodb!.NewImage) as Synonym
-      promises.push(putIndex(synonym))
+      promises.push(putIndex(group))
```
Comment (lpsinger) [inline app/table-streams/synonyms/index.ts:61]: You can simply the above as follows:

```ts
if (dynamoSynonyms.length > 0)
{
  await putIndex({
     synonymId,
     eventIds: dynamoSynonyms.map((synonym) => synonym.eventId),
 })
} else {
  await removeIndex(synonymId)
}
```

Then you can also remove the final `await Promises.all`.

**Review by lpsinger — CHANGES_REQUESTED** [review]

File: sandbox-search.js [hunk 7]
```diff
@@ -11,13 +11,23 @@ export default async function () {
   const text = await readFile('sandbox-seed.json', { encoding: 'utf-8' })
   const { circulars, synonyms } = JSON.parse(text)
 
+  const groupedSynonyms = synonyms.reduce((accumulator, synonym) => {
```
Comment (lpsinger) [inline sandbox-search.js:14]: I was thinking something more along the lines of https://lodash.com/docs/#groupBy.

**Review by lpsinger — CHANGES_REQUESTED** [review]

File: sandbox-search.js (hunk 7 above)
Comment (Courey) [reply sandbox-search.js:14]: is there a reason or is it just preference?

Comment (lpsinger) [reply sandbox-search.js:14]: It's terser and it's more obvious what the code does. I think it is more idiomatic. I would have suggested [`Object.groupBy`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/groupBy), but it is a relatively recent addition to ECMAScript and requires a newer version of Node.js than we are targeting.

File: sandbox-search.js [hunk 8]
```diff
@@ -6,16 +6,12 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 import { readFile } from 'fs/promises'
+import _ from 'lodash'
```
Comment (lpsinger) [inline sandbox-search.js:9]: Use the following import style to match what we do elsewhere in this project:

```suggestion
import groupBy from 'lodash/groupBy'
```

This is to facilitate tree-shaking. That doesn't matter in this context, but it's good to match the same style anyway.

**Review by lpsinger — CHANGES_REQUESTED** [review]

File: sandbox-search.js [hunk 9]
```diff
@@ -6,18 +6,24 @@
...
 
 export default async function () {
   const text = await readFile('sandbox-seed.json', { encoding: 'utf-8' })
   const { circulars, synonyms } = JSON.parse(text)
+  const groupedSynonyms = _.groupBy(synonyms, 'synonymId')
+  const groups = Object.keys(groupedSynonyms).map((id) => ({
+    synonymId: id,
+    eventIds: groupedSynonyms[id],
+  }))
```
Comment (lpsinger) [inline sandbox-search.js:18]:
```suggestion
  const groups = Object.entries(groupBy(synonyms, 'synonymId')).map(
    ([synonymId, values]) => [{synonymId, eventIds: values.map(({eventId}) => eventId)}])
```

File: sandbox-search.js [hunk 10]
```diff
@@ -6,18 +6,24 @@
...
 
   return [
     ...circulars.flatMap((item) => [
       { index: { _index: 'circulars', _id: item.circularId.toString() } },
       item,
     ]),
-    ...synonyms.flatMap((item) => [
-      { index: { _index: 'synonyms', _id: item.id } },
+    ...groups.flatMap((item) => [
```
Comment (lpsinger) [inline sandbox-search.js:25]:
```suggestion
    ...groups.map((item) => [
```

**Review by lpsinger — CHANGES_REQUESTED** [review]

**Review by lpsinger — APPROVED** [review]

**Pull request merged**